keras_cv/layers/preprocessing/aug_mix.py

```diff
@@ -0,0 +1,313 @@
+    chains of augmentations together using weights sampled from a Dirichlet
+    distribution.  A single chain consist of a series of individual augmentations.
+    The resultant image is further mixed with the original image to form the
+    final augmented image.
+
+    Args:
+        value_range: the range of values the incoming images will have.
+            Represented as a two number tuple written [low, high].
+            This is typically either `[0, 1]` or `[0, 255]` depending
+            on how your preprocessing pipeline is setup.
+        severity: A tuple of two floats, a single float or a `keras_cv.FactorSampler`.
+            A value is sampled from the provided range.  If a float is passed, the
+            range is interpreted as `(0, severity)`. This value represents the
+            level of strength of augmentations.  Defaults to 0.3.
+        width: an integer representing the number of different chains to
+            be mixed. Defaults to 3.
+        depth: an integer representing the number of transformations in
+            the chains. A negative value enables stochastic depth uniformly
+            in [1,3]. Defaults to -1.
+        alpha: a float value used as the probability coefficients for the
+            Beta and Dirichlet distributions. Defaults to 1.0.
+        seed: Integer. Used to create a random seed.
+
+    References:
+        [AugMix paper](https://arxiv.org/pdf/1912.02781)
+        [Official Code](https://github.com/google-research/augmix)
+        [Unoffial TF Code](https://github.com/szacho/augmix-tf)
+
+    Sample usage:
+    ```python
+    (images, labels), _ = tf.keras.datasets.cifar10.load_data()
+    augmix = keras_cv.layers.preprocessing.mix_up.AugMix([0, 255])
+    augmented_images = augmix(images)
+    ```
+    """
+
+    def __init__(
+        self,
+        value_range,
+        severity=0.3,
+        width=3,
+        depth=-1,
+        alpha=1.0,
+        seed=None,
+        **kwargs,
+    ):
+        super().__init__(seed=seed, **kwargs)
+        self.value_range = value_range
+        self.width = width
+        self.depth = depth
+        self.alpha = alpha
+        self.seed = seed
+        self.auto_vectorize = False
+        self.severity = severity
+
+        # initialize layers
+        self.auto_contrast = layers.AutoContrast
+        self.equalize = layers.Equalization
+        self.random_shear = layers.RandomShear
+
+    @staticmethod
+    def _sample_from_dirichlet(alpha):
+        gamma_sample = tf.random.gamma(shape=(), alpha=alpha)
+        return gamma_sample / tf.reduce_sum(gamma_sample, axis=-1, keepdims=True)
+
+    @staticmethod
+    def _sample_from_beta(alpha, beta):
+        sample_alpha = tf.random.gamma((), 1.0, beta=alpha)
+        sample_beta = tf.random.gamma((), 1.0, beta=beta)
+        return sample_alpha / (sample_alpha + sample_beta)
+
+    def _sample_level(self, level, maxval, dtype):
+        level = (
+            preprocessing.parse_factor(
```

Comment:
we don't want to parse a factor every call to `_sample_level()`

Is level always a single number in these calls?  Is it dynamic in that there are many calls?

If so, we may need to use a different abstraction here (maybe just directly compute the samples instead - similar to _sample_from_beta and _sample_from_dirichlet).

We definitely dont want to parse a factor every invocation.

Reply:
Yes, I believe we should do this. Levels are calculated dynamically.

I think initially this wasn't implemented using parse_factor and this was due to a review comment. I will simply revert back!

Reply:
I imagine what what I mean't to ask with the comment is whether or not we use the factor API in the constructor.  If it is not possible, that is totally fine.

Reply:
Yeah, I don't think it's possible. We need different level values for each call.